components/butterfly/src/rumor.rs

```diff
@@ -333,6 +388,46 @@ mod storage {
             }
             result
         }
+
+        pub fn keys_rsr(&self) -> Vec<RumorKey> {
+            self.list
+                .read()
+                .values()
+                .flat_map(HashMap::values)
+                .map(RumorKey::from)
+                .collect()
+        }
+    }
+
+    impl<T: Transient> RumorStore<T> {
+        pub fn insert_transient_rsw(&self, rumor: T) -> bool {
+            if !rumor.expired(Utc::now()) {
+                self.insert_rsw(rumor)
+            } else {
+                false
+            }
+        }
+
+        /// Remove all rumors that have expired from our rumor store.
+        pub fn purge_expired_rsr(&self, expiration_date: DateTime<Utc>) {
+            self.lock_rsr()
+                .rumors()
+                .filter(|r| r.expiration().expired(expiration_date))
+                .for_each(|r| {
+                    self.remove_rsw(r.key(), r.id());
+                });
+        }
+
+        pub fn with_rumors_transient_rsw<F>(&self, key: &str, closure: F)
+            where F: FnMut(&mut T)
```

Comment:
Similar to `purge_expired_rsr`, I think this can be replaced with a mutable version of `ServiceGroupRumors`. Something like:
```rust
    impl<'sg_key, R> ServiceGroupRumorsMut<'sg_key, R> {
        pub fn map_rumor<OUT>(&mut self,
                              member_id: &str,
                              f: impl Fn(&mut R) -> OUT)
                              -> Option<OUT> {
            self.0.as_mut().and_then(|m| m.get_mut(member_id).map(f))
        }

        pub fn rumors(&'sg_key mut self) -> impl Iterator<Item = &mut R> {
            self.0.iter_mut().map(|m| m.values_mut()).flatten()
        }
    }
```
should allow you to replace the callers with:
```rust
        self.service_store
            .lock_rsw()
            .service_group(&service_group)
            .map_rumor(&self.member_id, Transient::refresh);
```
and
```rust
        self.service_config_store
            .lock_rsw()
            .service_group(&service_group)
            .rumors()
            .for_each(Transient::refresh);
```

Note that this doesn't even require a `Transient` bound.

Reply:
Sounds good